File: app/src/main/java/com/joshuahalvorson/safeyoutube/view/activity/WatchPlaylistActivity.java
package com.joshuahalvorson.safeyoutube.view.activity;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;
import com.google.android.youtube.player.YouTubeInitializationResult;
import com.google.android.youtube.player.YouTubePlayer;
import com.google.android.youtube.player.YouTubePlayerSupportFragment;
import com.joshuahalvorson.safeyoutube.ApiKey;
import com.joshuahalvorson.safeyoutube.R;
import com.joshuahalvorson.safeyoutube.adapter.PlaylistItemsListRecyclerviewAdapter;
import com.joshuahalvorson.safeyoutube.model.Item;
import com.joshuahalvorson.safeyoutube.model.PlaylistResultOverview;
import com.joshuahalvorson.safeyoutube.network.YoutubeDataApiViewModel;

import java.util.ArrayList;

public class WatchPlaylistActivity extends AppCompatActivity {
    public static final String PLAYLIST_ID_KEY = "playlistId";
    private YouTubePlayerSupportFragment youTubePlayerFragment;
    private YoutubeDataApiViewModel viewModel;
    private ArrayList<Item> items;
    private PlaylistItemsListRecyclerviewAdapter adapter;
    private String playlistId;
    private SharedPreferences sharedPref;
    private int ageValue;
    private YouTubePlayer mYoutubePlayer;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_watch_playlist);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        items = new ArrayList<>();
        adapter = new PlaylistItemsListRecyclerviewAdapter(items, new PlaylistItemsListRecyclerviewAdapter.OnVideoClicked() {
            @Override
            public void onVideoClicked(int itemIndex) {
                mYoutubePlayer.loadPlaylist(playlistId, itemIndex, 1);
                mYoutubePlayer.play();
            }
        });

        viewModel = ViewModelProviders.of(this).get(YoutubeDataApiViewModel.class);
        youTubePlayerFragment =
                (YouTubePlayerSupportFragment) getSupportFragmentManager()
                        .findFragmentById(R.id.youtube_fragment);
        RecyclerView videosRecyclerview = findViewById(R.id.videos_list);
        videosRecyclerview.setLayoutManager(new LinearLayoutManager(getApplicationContext()));
        videosRecyclerview.setAdapter(adapter);
        if(getIntent() != null) {
            playlistId = getIntent().getStringExtra(PLAYLIST_ID_KEY);
            initializeVideo(youTubePlayerFragment, playlistId, 0);
        }

        LiveData<PlaylistResultOverview> playlistResultOverview =
                viewModel.getPlaylistOverview(playlistId);

        playlistResultOverview.observe(this, new Observer<PlaylistResultOverview>() {
            @Override
            public void onChanged(@Nullable PlaylistResultOverview playlistResultOverview) {
                items.clear();
                items.addAll(playlistResultOverview.getItems());
                adapter.notifyDataSetChanged();
            }
        });

    }

    @Override
    protected void onResume() {
        super.onResume();
        sharedPref = getSharedPreferences(
                getString(R.string.preference_file_key), Context.MODE_PRIVATE);
        ageValue = sharedPref.getInt(getString(R.string.age_range_key), 1);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_settings) {
            Intent loginIntent = new Intent(getApplicationContext(), LoginActivity.class);
            startActivityForResult(loginIntent, SettingsActivity.LOGIN_REQUEST_CODE);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void initializeVideo(YouTubePlayerSupportFragment fragment, final String playlistId, final int itemIndex){
        fragment.initialize(ApiKey.KEY, new YouTubePlayer.OnInitializedListener() {
            @Override
            public void onInitializationSuccess(YouTubePlayer.Provider provider, YouTubePlayer youTubePlayer, boolean b) {
                mYoutubePlayer = youTubePlayer;
                mYoutubePlayer.loadPlaylist(playlistId);
                switch (ageValue){
                    case 0:
                        mYoutubePlayer.setPlayerStyle(YouTubePlayer.PlayerStyle.CHROMELESS);
                        break;
                    case 1:
                        mYoutubePlayer.setPlayerStyle(YouTubePlayer.PlayerStyle.MINIMAL);
                        break;
                    case 2:
                        mYoutubePlayer.setPlayerStyle(YouTubePlayer.PlayerStyle.DEFAULT);
                        break;
                }

            }

            @Override
            public void onInitializationFailure(YouTubePlayer.Provider provider, YouTubeInitializationResult youTubeInitializationResult) {
                if (youTubeInitializationResult.isUserRecoverableError()) {
                    youTubeInitializationResult.getErrorDialog(getParent(), 1).show();
                } else {
                    String errorMessage = String.format(
                            getString(R.string.error_player), youTubeInitializationResult.toString());
                    Toast.makeText(getApplicationContext(), errorMessage, Toast.LENGTH_LONG).show();
                }
            }
        });
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if(requestCode == SettingsActivity.LOGIN_REQUEST_CODE && resultCode == RESULT_OK){
            Log.i("login", "logged in");
            Intent intent = new Intent(getApplicationContext(), SettingsActivity.class);
            startActivity(intent);
        }else if (requestCode == SettingsActivity.LOGIN_REQUEST_CODE && resultCode == RESULT_CANCELED){
            Log.i("login", "not logged in");
            Snackbar.make(findViewById(R.id.videos_list), "Not logged in", Snackbar.LENGTH_SHORT).show();
        }
    }

}

File: app/src/main/java/com/joshuahalvorson/safeyoutube/view/activity/WatchPlaylistActivity.kt
package com.joshuahalvorson.safeyoutube.view.activity

import android.annotation.SuppressLint
import android.content.Context
import android.content.DialogInterface
import android.content.SharedPreferences
import android.content.pm.ActivityInfo
import android.content.res.Configuration
import android.hardware.SensorManager
import android.os.Bundle
import android.util.Log
import android.view.OrientationEventListener
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.joshuahalvorson.safeyoutube.R
import com.joshuahalvorson.safeyoutube.adapter.ItemsRecyclerviewAdapter
import com.joshuahalvorson.safeyoutube.model.Models
import com.joshuahalvorson.safeyoutube.network.YoutubeDataApiViewModel
import com.joshuahalvorson.safeyoutube.util.Counter
import com.joshuahalvorson.safeyoutube.view.fragment.LoginFragment
import com.joshuahalvorson.safeyoutube.util.YoutubePlayerController
import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.PlayerConstants
import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.PlayerConstants.PlayerState.ENDED
import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.YouTubePlayer
import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.listeners.AbstractYouTubePlayerListener
import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.listeners.YouTubePlayerCallback
import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.listeners.YouTubePlayerFullScreenListener
import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.views.YouTubePlayerView
import com.pierfrancescosoffritti.androidyoutubeplayer.core.ui.PlayerUiController
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_watch_playlist.*

class WatchPlaylistActivity : AppCompatActivity(), DialogInterface.OnDismissListener {
    override fun onDismiss(dialog: DialogInterface?) {
        hideSystemUI()
    }

    private var items: ArrayList<Models.Item> = ArrayList()
    private var ageValue: Int? = 0

    private lateinit var itemAdapter: ItemsRecyclerviewAdapter
    private lateinit var sharedPref: SharedPreferences
    private lateinit var counter: Counter
    private lateinit var playerController: YoutubePlayerController
    private lateinit var uiController: PlayerUiController
    private lateinit var viewModel: YoutubeDataApiViewModel
    private lateinit var disposable: CompositeDisposable

    @SuppressLint("CheckResult")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_watch_playlist)

        disposable = CompositeDisposable()

        sharedPref = getSharedPreferences(
                getString(R.string.preference_file_key), Context.MODE_PRIVATE)

        viewModel = ViewModelProviders.of(this).get(YoutubeDataApiViewModel::class.java)

        val youtubePlayerView = findViewById<YouTubePlayerView>(R.id.youtube_player_view)
        lifecycle.addObserver(youtubePlayerView)
        uiController = youtubePlayerView.getPlayerUiController()
        uiController.showMenuButton(false)
        uiController.showYouTubeButton(false)
        youtubePlayerView.addFullScreenListener(object : YouTubePlayerFullScreenListener {
            override fun onYouTubePlayerEnterFullScreen() {
                settings_fab.hide()
            }

            override fun onYouTubePlayerExitFullScreen() {
                settings_fab.show()
            }
        })

        settings_fab.setOnClickListener {
            LoginFragment().show(supportFragmentManager, "login_fragment")
        }

        itemAdapter = ItemsRecyclerviewAdapter(items, object : ItemsRecyclerviewAdapter.OnVideoClicked {
            override fun onVideoClicked(itemIndex: Int) {
                counter = Counter(0, itemIndex, items.size - 1)
                items[itemIndex].contentDetails?.videoId?.let {
                    playerController.playVideo(it)
                }
            }
        })

        videos_list.apply {
            videos_list.layoutManager = LinearLayoutManager(applicationContext)
            videos_list.adapter = itemAdapter
        }

        val orientationEventListener = object : OrientationEventListener(applicationContext, SensorManager.SENSOR_DELAY_UI) {
            override fun onOrientationChanged(orientation: Int) {
                if (orientation <= 10) {
                    requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED
                }
            }
        }

        if (orientationEventListener.canDetectOrientation()) {
            orientationEventListener.enable()
        }
    }

    override fun onResume() {
        super.onResume()
        hideSystemUI()
        val currentPlaylistId = sharedPref.getString(getString(R.string.current_playlist_key), null)
        if (currentPlaylistId != null) {
            viewModel.getPlaylistOverview(currentPlaylistId)
                    ?.subscribeOn(Schedulers.io())
                    ?.observeOn(AndroidSchedulers.mainThread())
                    ?.subscribe {
                        Log.i("asda", it.items[0].snippet?.title)
                        items.clear()
                        items.addAll(it.items)
                        counter = Counter(0, 0, items.size - 1)
                        itemAdapter.notifyDataSetChanged()
                        no_playlist_text.visibility = View.GONE
                        youtube_player_view.getYouTubePlayerWhenReady(object : YouTubePlayerCallback {
                            override fun onYouTubePlayer(youTubePlayer: YouTubePlayer) {
                                counter = Counter(0, 0, items.size - 1)
                                youtube_player_view.visibility = View.VISIBLE
                                items[0].contentDetails?.videoId?.let { youTubePlayer.cueVideo(it, 0F) }
                                playerController = YoutubePlayerController(youTubePlayer)
                                initializeYoutubePlayer(youtube_player_view)
                            }
                        })
                    }?.let { subscribe ->
                        disposable.add(
                                subscribe
                        )
                    }

        } else {
            items.clear()
            youtube_player_view.visibility = View.GONE
            no_playlist_text.visibility = View.VISIBLE
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        disposable.dispose()
    }

    private fun initializeYoutubePlayer(youtubePlayer: YouTubePlayerView) {
        youtubePlayer.addYouTubePlayerListener(object : AbstractYouTubePlayerListener() {
            override fun onStateChange(youTubePlayer: YouTubePlayer, state: PlayerConstants.PlayerState) {
                when (state) {
                    ENDED -> {
                        youTubePlayer.loadVideo(items[counter.increment()].contentDetails?.videoId.toString(), 0F)
                    }
                    else -> return
                }
            }

            override fun onError(youTubePlayer: YouTubePlayer, error: PlayerConstants.PlayerError) {
                Toast.makeText(applicationContext, "Error is: $error", Toast.LENGTH_LONG).show()
            }
        })
    }

    override fun onBackPressed() {
        if (resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE) {
            requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
        } else {
            super.onBackPressed()
        }
    }

    override fun onConfigurationChanged(newConfig: Configuration?) {
        super.onConfigurationChanged(newConfig)
        if (newConfig?.orientation == Configuration.ORIENTATION_LANDSCAPE) {
            youtube_player_view.enterFullScreen()
            uiController.showFullscreenButton(false)
        } else {
            youtube_player_view.exitFullScreen()
            uiController.showFullscreenButton(true)
        }
    }

    private fun hideSystemUI() {
        window.decorView.systemUiVisibility = (View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                or View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                or View.SYSTEM_UI_FLAG_FULLSCREEN)
    }

    override fun onStop() {
        super.onStop()
        sharedPref.edit().remove(getString(R.string.current_playlist_key)).apply()
    }
}
